osu.Game/Overlays/UserProfileOverlay.cs

```diff
@@ -107,37 +117,69 @@ public void ShowUser(User user, bool fetchOnline = true)
 
             if (fetchOnline)
             {
-                userReq = new GetUserRequest(user.Id);
-                userReq.Success += userLoadComplete;
-                API.Queue(userReq);
+                fetchOnlineUser(user.Id);
             }
             else
             {
-                userReq = null;
                 userLoadComplete(user);
             }
 
             sectionsContainer.ScrollToTop();
         }
 
-        private void userLoadComplete(User user)
+        private void onRulesetChanged(RulesetInfo ruleset)
         {
-            Header.User.Value = user;
+            if (displayedUser.Value != null)
+            {
+                Header.Loading = true;
+                fetchOnlineUser(displayedUser.Value.Id, ruleset);
+            }
+        }
+
+        private void fetchOnlineUser(long userId, RulesetInfo ruleset = null)
+        {
+            userReq?.Cancel();
 
-            if (user.ProfileOrder != null)
+            userReq = new GetUserRequest(userId, ruleset);
+            userReq.Success += userLoadComplete;
+            API.Queue(userReq);
+        }
+
+        private void userLoadComplete(User user)
+        {
+            // If new user has been loaded
+            if (displayedUser.Value?.Id != user.Id)
             {
-                foreach (string id in user.ProfileOrder)
+                if (user.ProfileOrder != null)
                 {
-                    var sec = sections.FirstOrDefault(s => s.Identifier == id);
-
-                    if (sec != null)
+                    foreach (string id in user.ProfileOrder)
                     {
-                        sec.User.Value = user;
+                        var sec = sections.FirstOrDefault(s => s.Identifier == id);
+
+                        if (sec != null)
+                        {
+                            sec.User.Value = user;
 
-                        sectionsContainer.Add(sec);
-                        tabs.AddItem(sec);
+                            sectionsContainer.Add(sec);
+                            tabs.AddItem(sec);
+                        }
                     }
                 }
+
+                // Allows change without triggering
+                displayedRuleset.UnbindAll();
```

Comment:
this is near impossible to follow logic-wise. unbinding one bindable, propagating another then rebinding is not something we should be doing.

Reply:
To elaborate on this we generally fix these kinds of scenarios by interrupting the bindable flow at the point of transfer. So either `RulesetSelector` or `ProfileHeader would expose a public ruleset bindable (for the overlay to bind to) but also hold a private bindable. Then you can control the ValueChanged being fired on the outer one (which would cause a re-query) and avoid it happening for the case you are trying to avoid here.

You can see an example of this in SongSelect (ruleset / beatmap). Let me know if this doesn't make sense.